File: android/lib/src/main/java/Beatmup/Android/Bitmap.java
/*
    Beatmup image and signal processing library
    Copyright (C) 2019, lnstadrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package Beatmup.Android;

import android.graphics.BitmapFactory;

import java.io.FileDescriptor;
import java.io.InputStream;

import Beatmup.Context;

/**
 * Android bitmap wrapping.
 * Enables access to Android bitmaps from inside the Beatmup engine without memory copy.
 */
public class Bitmap extends Beatmup.Bitmap {
    private android.graphics.Bitmap source;     //!< wrapped Android bitmap

    /**
     * Creates new bitmap from Android bitmap object without memory copy.
     * @param context   Beatmup context
     * @param bitmap    source bitmap
     */
    private Bitmap(Context context, android.graphics.Bitmap bitmap) {
        super(context, bitmap);
        source = bitmap;
    }

    /**
     * Creates new bitmap from a source Android bitmap converting the data into appropriate pixel format if necessary.
     * @param context   the context handling the new bitmap
     * @param bitmap    the source; may be recycled
     * @return new bitmap
     */
    private static Bitmap createEnsuringPixelFormat(Context context, android.graphics.Bitmap bitmap) {
        if (bitmap.getConfig() != android.graphics.Bitmap.Config.ALPHA_8  &&  bitmap.getConfig() != android.graphics.Bitmap.Config.ARGB_8888) {
            android.graphics.Bitmap copy = bitmap.copy(android.graphics.Bitmap.Config.ARGB_8888, true);
            bitmap.recycle();
            return new Bitmap(context, copy);
        } else
            return new Bitmap(context, bitmap);
    }

    /**
     * Decodes a bitmap from stream.
     * @param context       a Beatmup context
     * @param inputStream   the stream to decode
     * @return bitmap containing the decoded stream content
     */
    public static Bitmap decodeStream(Context context, InputStream inputStream) throws OutOfMemoryError {
        android.graphics.Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
        if (bitmap == null)
            return null;
        return createEnsuringPixelFormat(context, bitmap);
    }

    /**
     * Loads a bitmap from file. If the pixel format is not supported, this function tries to convert the source data to a 32-bit color bitmap.
     * @param context       a Beatmup context
     * @param path          path to the file to decode
     * @param options       bitmap loader options
     * @return bitmap from the file
     */
    public static Bitmap decodeFile(Context context, String path, BitmapFactory.Options options) throws OutOfMemoryError {
        android.graphics.Bitmap bitmap = BitmapFactory.decodeFile(path, options);
        if (bitmap == null)
            return null;
        return createEnsuringPixelFormat(context, bitmap);
    }

    /**
     * Loads a bitmap from file.
     * @param context       a Beatmup context
     * @param file          a file descriptor of the content to decode
     * @param outPadding    outPadding returned by BitmapFactory.decodeFileDescriptor
     * @param options       decoding options
     * @return bitmap from the file
     */
    public static Bitmap decodeFileDescriptor(
            Context context,
            FileDescriptor file,
            android.graphics.Rect outPadding,
            BitmapFactory.Options options
    ) throws OutOfMemoryError
    {
        android.graphics.Bitmap bitmap = BitmapFactory.decodeFileDescriptor(file, outPadding, options);
        if (bitmap == null)
            return null;
        return createEnsuringPixelFormat(context, bitmap);
    }

    /**
     * Creates empty ARGB bitmap of specified size
     * @param context   a Beatmup context
     * @param width     width of the new bitmap in pixels
     * @param height    height of the new bitmap in pixels
     * @return the new bitmap
     */
    public static Bitmap createColorBitmap(Context context, int width, int height) {
        return new Bitmap(context, android.graphics.Bitmap.createBitmap(width, height, android.graphics.Bitmap.Config.ARGB_8888));
    }

    /**
     * Creates empty grayscale bitmap of specified size
     * @param context   a Beatmup context
     * @param width     width of the new bitmap in pixels
     * @param height    height of the new bitmap in pixels
     * @return the new bitmap
     */
    public static Bitmap createGrayscaleBitmap(Context context, int width, int height) {
        return new Bitmap(context, android.graphics.Bitmap.createBitmap(width, height, android.graphics.Bitmap.Config.ALPHA_8));
    }

    /**
     * @return the wrapped Android bitmap instance.
     */
    public android.graphics.Bitmap getBitmap() {
        return source;
    }

    @Override
    public synchronized void dispose() {
        if (source != null) {
            source.recycle();
            source = null;
        }
        super.dispose();
    }

    /**
     * @return copy of this bitmap
     */
    public Bitmap clone() {
        return new Bitmap(context, source.copy(source.getConfig(), true));
    }
}

File: python/src/bitmap.cpp
/*
    Beatmup image and signal processing library
    Copyright (C) 2020, lnstadrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "bitmap.h"
#include <stdexcept>

using namespace Beatmup;


Python::Bitmap::Bitmap(Beatmup::Context& context, pybind11::buffer& buffer):
    AbstractBitmap(context)
{
    auto info = buffer.request();

    // check if empty
    if (info.size == 0)
        throw std::invalid_argument("A bitmap cannot be empty");

    // check dimensions
    if (info.ndim < 3)
        throw std::invalid_argument("A bitmap is expected to have at least 3 dimensions");
    else
        for (int i = 0; i + 3 < info.ndim; ++i)
            if (info.shape[i] != 1)
                throw std::invalid_argument("A bitmap is expected to have at most 3 inner non-singleton dimensions.");
    
    // check pixel format
    static const char* FORMAT_UINT8 = "B";
    static const char* FORMAT_FLOAT = "f";
    const auto numChannels = info.shape[info.ndim - 1];
    
    if (info.format == FORMAT_UINT8)
        switch (numChannels) {
            case 1: format = PixelFormat::SingleByte; break;
            case 3: format = PixelFormat::TripleByte; break;
            case 4: format = PixelFormat::QuadByte; break;
            default:
                throw std::invalid_argument("A bitmap may have 1, 3 of 4 channels, but got " + std::to_string(numChannels));
        }
    else if (info.format == FORMAT_FLOAT)
        switch (numChannels) {
            case 1: format = PixelFormat::SingleFloat; break;
            case 3: format = PixelFormat::TripleFloat; break;
            case 4: format = PixelFormat::QuadFloat; break;
            default:
                throw std::invalid_argument("A bitmap may have 1, 3 of 4 channels, but got " + std::to_string(numChannels));
        }
    else
        throw std::invalid_argument("Unsupported data format. Expected bitmap values of type float32 or uint8.");

    // check stride
    const int bps = getBitsPerPixel() / 8;
    strideBytes = info.shape[info.ndim - 2] * bps;
    if (info.strides[info.ndim - 1] * numChannels != bps || info.strides[info.ndim - 2] != bps || info.strides[info.ndim - 3] != strideBytes)
        throw std::invalid_argument("Strided bitmaps are not supported.");

    // GOOOD.
    data = std::move(info);
}


const pixbyte* Python::Bitmap::getData(int x, int y) const { 
    return static_cast<const pixbyte*>(data.ptr) + y * strideBytes + x;
}


pixbyte* Python::Bitmap::getData(int x, int y) { 
    return static_cast<pixbyte*>(data.ptr) + y * strideBytes + x;
}


const msize Python::Bitmap::getMemorySize() const { 
    return data.itemsize * data.size;
}